File: NEW CODES/LP-II/AI/Code/Prac 1(BFS And DFS)/Python/BFSandDFS.py
from collections import deque


class Graph:
    def __init__(self, v):
        self.V = v
        self.adj = [[] for _ in range(v)]

    def addEdge(self, v, w):
        self.adj[v].append(w)

    def DFSUtil(self, vertex, visited):
        visited[vertex] = True
        print(vertex, end=" ")

        for a in self.adj[vertex]:
            if not visited[a]:
                self.DFSUtil(a, visited)

    def DFS(self, v):
        visited = [False] * self.V
        self.DFSUtil(v, visited)

    def BFS(self, s):
        visited = [False] * self.V
        queue = deque()
        visited[s] = True
        queue.append(s)

        while queue:
            s = queue.popleft()
            print(s, end=" ")

            for n in self.adj[s]:
                if not visited[n]:
                    visited[n] = True
                    queue.append(n)


n = int(input("Enter the size of the graph: "))
g = Graph(n)

size = int(input("Enter the size of input: "))
for i in range(size):
    j, k = map(int, input(f"Enter edges {i + 1} of graph: ").split())

    if j < n and k < n:
        g.addEdge(j, k)
    else:
        print("Invalid Input")

start = int(input("Enter the starting vertex: "))
print("DFS of Graph")
g.DFS(start)
print()
print("BFS of Graph")
g.BFS(start)

"""
Enter the size of the graph: 5  
Enter the size of input: 6  
Enter edges 1 of graph: 0 1
Enter edges 2 of graph: 0 2
Enter edges 3 of graph: 0 3
Enter edges 4 of graph: 1 2
Enter edges 5 of graph: 2 4
Enter edges 6 of graph: 3 4  
Enter the starting vertex:0
BFS of Graph  0 1 2 3 4
DFS of Graph  0 1 2 4 3
"""

File: NEW CODES/LP-II/AI/Main_Code/Prac 1(BFS And DFS)/Java/BFSandDFS.java
import java.util.*;

class Graph
{   private int V; 
    private LinkedList<Integer> adj[];  
    public Graph(int v)
    {   
        V=v;
        adj = new LinkedList[v];
        for (int i = 0; i < v; ++i)
        {
          adj[i] = new LinkedList();
    	}
    }        
    void addEdge(int v, int w)
    {
        adj[v].add(w);                              
    }
    void DFSUtil(int vertex, boolean visit[])
    {

        visit[vertex] = true;                         
        System.out.print(vertex + " ");
        int a = 0;
        for (int i = 0; i < adj[vertex].size(); i++)  
            {
                a = adj[vertex].get(i);
                if (!visit[a])                   
                {
                    DFSUtil(a, visit);
                }
            }  
    }

    void DFS(int v)
    {
        boolean already[] = new boolean[V];             
        DFSUtil(v, already);
    }

    void BFS(int s) {
        boolean visited[] = new boolean[V];
        BFSUtil(s, visited);
    }

    void BFSUtil(int s, boolean visited[]) {
        LinkedList<Integer> queue = new LinkedList<Integer>();
        visited[s] = true;
        queue.add(s);

        while (queue.size() != 0) {
            s = queue.poll();
            System.out.print(s + " ");

            Iterator<Integer> i = adj[s].listIterator();
            while (i.hasNext()) {
                int n = i.next();
                if (!visited[n]) {
                    visited[n] = true;
                    queue.add(n);
                }
            }
        }
    }

}
 
class Main
{
    public static void main(String[] args)
    {
        Scanner in= new Scanner(System.in);
        System.out.print("Enter the size of the graph: ");
        int n = in.nextInt();
        System.out.print("Enter the size of input: ");
        int size=in.nextInt();
        Graph g=new Graph(n);
        for(int i=0;i<size;i++){
            System.out.print("Enter edges  "+(i+1)+ " of graph: ");
            int j=in.nextInt();
            int k=in.nextInt();
            if(j<n && k<n){
               g.addEdge(j, k);
            }
            else{
                System.out.println("Invalid Input");
            }
        }
        System.out.println("Enter the starting vertex: ");
        int start=in.nextInt();
        System.out.println("DFS of Graph");
        g.DFS(start);
        System.out.println();
        System.out.println("BFS of Graph");
        g.BFS(start);
    }
}

/*
Enter the size of the graph: 5  
Enter the size of input: 6  
Enter edges 1 of graph: 0 1
Enter edges 2 of graph: 0 2
Enter edges 3 of graph: 0 3
Enter edges 4 of graph: 1 2
Enter edges 5 of graph: 2 4
Enter edges 6 of graph: 3 4  
Enter the starting vertex:0
BFS of Graph  0 1 2 3 4
DFS of Graph  0 1 2 4 3
*/